File: 153.find-minimum-in-rotated-sorted-array.py
#
# @lc app=leetcode id=153 lang=python3
#
# [153] Find Minimum in Rotated Sorted Array
#


# @lc code=start
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # binary search
        # only two senario of mid point
        # if nums[mid-1]>nums[mid] && nums[mid+1]>nums[mid], return ans
        # else:
        #   if nums[r]<nums[mid], the answer must in upper half
        #   else: lower half
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if (nums[mid - 1] > nums[mid] if mid - 1 >= 0 else True) and (
                nums[mid + 1] > nums[mid] if mid + 1 < len(nums) else True
            ):
                return nums[mid]
            else:
                if nums[r] < nums[mid]:
                    l = mid + 1
                else:
                    r = mid - 1


# @lc code=end

File: leetcode-nvim/153.find-minimum-in-rotated-sorted-array.go
package main

// @leet start
func findMin(nums []int) int {
	// binary search
	// if nums[m-1] > nums[m] && nums[m+1] > nums[m] --> Find answer
	// if nums[r] < nums[m] --> Find answer in upper half
	// else, lower hald
	l, r := 0, len(nums)-1
	n := len(nums)
	for l <= r {
		m := (l + r) / 2
		lowerThanRightNum := false
		if m+1 < n {
			lowerThanRightNum = nums[m] < nums[m+1]
		} else {
			lowerThanRightNum = true
		}
		lowerThanLeftNum := false
		if m-1 >= 0 {
			lowerThanLeftNum = nums[m] < nums[m-1]
		} else {
			lowerThanLeftNum = true
		}
		if lowerThanRightNum && lowerThanLeftNum {
			return nums[m]
		}
		if nums[m] > nums[r] {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	// should never reach here
	return nums[l]
}

// @leet end
